File: src/funcs/apis/tidalapi.py
import datetime
import json
import argparse
import requests
import os
from base64 import b64encode
import sys
class TidalAPI:
    API_URL = "https://openapi.tidal.com/v2"
    TOKEN_URL = "https://auth.tidal.com/v1/oauth2/token"
    STREAM_API_URL = "https://api.tidal.com/v1"  # Different base URL for stream endpoint

    def __init__(self):
        self.session = requests.Session()
        self.access_token = None
        self.token_expiry = None
        self.load_credentials()
        self.authenticate()

    def load_credentials(self):
        try:
            # First check if apis.json exists in the current directory
            if os.path.exists('apis.json'):
                config_path = 'apis.json'
            else:
                # Try to find it in the script's directory
                script_dir = os.path.dirname(os.path.abspath(__file__))
                config_path = os.path.join(script_dir, 'apis.json')

                if not os.path.exists(config_path):
                    raise FileNotFoundError("apis.json not found")

            with open(config_path, 'r') as f:
                credentials = json.load(f)

            if 'TIDAL_CLIENT_ID' not in credentials or 'TIDAL_CLIENT_SECRET' not in credentials:
                raise KeyError("TIDAL_CLIENT_ID or TIDAL_CLIENT_SECRET not found in apis.json")

            self.client_id = credentials['TIDAL_CLIENT_ID']
            self.client_secret = credentials['TIDAL_CLIENT_SECRET']

        except Exception as e:
            print(json.dumps({'error': f"Failed to load credentials: {str(e)}"}, indent=2))
            sys.exit(1)
    def authenticate(self):
        if self.access_token and self.token_expiry and datetime.datetime.now() < self.token_expiry:
            return
        auth_header = b64encode(f"{self.client_id}:{self.client_secret}".encode()).decode()
        headers = {
            "Authorization": f"Basic {auth_header}",
            "Content-Type": "application/x-www-form-urlencoded"
        }
        data = {"grant_type": "client_credentials"}

        response = self.session.post(self.TOKEN_URL, headers=headers, data=data)
        response.raise_for_status()
        token_data = response.json()
        self.access_token = token_data['access_token']
        self.token_expiry = datetime.datetime.now() + datetime.timedelta(seconds=token_data['expires_in'])

    def make_request(self, method, url, params=None, base_url=None):
        self.authenticate()
        headers = {
            "Authorization": f"Bearer {self.access_token}",
            "Accept": "application/vnd.api+json"
        }
        full_url = f"{base_url or self.API_URL}/{url}"
        response = self.session.request(method, full_url, headers=headers, params=params)
        response.raise_for_status()
        return response.json()

    def get_stream_url(self, track_id: str, country_code: str, user_id: str = None, user_token: str = None):
        """Get the stream URL for a track using either client credentials or user token."""
        headers = {
            "Authorization": f"Bearer {user_token if user_token else self.access_token}"
        }

        params = {'countryCode': country_code}
        url = f"{self.STREAM_API_URL}/tracks/{track_id}/streamUrl"

        response = self.session.get(url, headers=headers, params=params)
        response.raise_for_status()
        return response.json()

    def get_track(self, track_id: str, country_code: str):
        params = {"countryCode": country_code, "include": "artists,albums"}
        return self.make_request("GET", f"tracks/{track_id}", params=params)

    def get_album(self, album_id: str, country_code: str):
        # Fetch album details
        album_response = self.make_request(
            "GET",
            f"albums/{album_id}",
            params={"countryCode": country_code, "include": "items"}
        )

        # Check if album_response contains artist details or artistId
        if album_response and "artist" not in album_response:
            # Fetch artist details if needed
            artist_id = album_response.get("artistId")
            if artist_id:
                artist_response = self.make_request(
                    "GET",
                    f"artists/{artist_id}",
                    params={"countryCode": country_code}
                )
                # Add artist name to the album response
                if artist_response:
                    album_response["artistName"] = artist_response.get("name", "Unknown Artist")

        return album_response


    def search_tracks(self, query: str, country_code: str, limit: int = 30):
        self.authenticate()
        url = "https://openapi.tidal.com/search"
        params = {
            "query": query,
            "type": "TRACKS",
            "offset": 0,
            "limit": limit,
            "countryCode": country_code,
            "popularity": "WORLDWIDE"
        }
        headers = {
            "Authorization": f"Bearer {self.access_token}",
            "Content-Type": "application/vnd.tidal.v1+json",
            "Accept": "application/vnd.tidal.v1+json"
        }

        response = self.session.get(url, headers=headers, params=params)
        response.raise_for_status()
        return response.json()

    def search_albums(self, query: str, country_code: str, limit: int = 30):
        self.authenticate()
        url = "https://openapi.tidal.com/search"
        params = {
            "query": query,
            "type": "ALBUMS",
            "offset": 0,
            "limit": limit,
            "countryCode": country_code,
            "popularity": "WORLDWIDE"
        }
        headers = {
            "Authorization": f"Bearer {self.access_token}",
            "Content-Type": "application/vnd.tidal.v1+json",
            "Accept": "application/vnd.tidal.v1+json"
        }

        response = self.session.get(url, headers=headers, params=params)
        response.raise_for_status()
        return response.json()

    def search_artists(self, query: str, country_code: str, limit: int = 30):
        self.authenticate()
        url = "https://openapi.tidal.com/search"
        params = {
            "query": query,
            "type": "ARTISTS",
            "offset": 0,
            "limit": limit,
            "countryCode": country_code,
            "popularity": "WORLDWIDE"
        }
        headers = {
            "Authorization": f"Bearer {self.access_token}",
            "Content-Type": "application/vnd.tidal.v1+json",
            "Accept": "application/vnd.tidal.v1+json"
        }

        response = self.session.get(url, headers=headers, params=params)
        response.raise_for_status()
        return response.json()

    def search_playlists(self, query: str, country_code: str, limit: int = 30):
        self.authenticate()
        url = "https://openapi.tidal.com/search"
        params = {
            "query": query,
            "type": "PLAYLISTS",
            "offset": 0,
            "limit": limit,
            "countryCode": country_code,
            "popularity": "WORLDWIDE"
        }
        headers = {
            "Authorization": f"Bearer {self.access_token}",
            "Content-Type": "application/vnd.tidal.v1+json",
            "Accept": "application/vnd.tidal.v1+json"
        }

        response = self.session.get(url, headers=headers, params=params)
        response.raise_for_status()
        return response.json()

    def search_videos(self, query: str, country_code: str, limit: int = 30):
        self.authenticate()
        url = "https://openapi.tidal.com/search"
        params = {
            "query": query,
            "type": "VIDEOS",
            "offset": 0,
            "limit": limit,
            "countryCode": country_code,
            "popularity": "WORLDWIDE"
        }
        headers = {
            "Authorization": f"Bearer {self.access_token}",
            "Content-Type": "application/vnd.tidal.v1+json",
            "Accept": "application/vnd.tidal.v1+json"
        }

        response = self.session.get(url, headers=headers, params=params)
        response.raise_for_status()
        return response.json()

def main():
    parser = argparse.ArgumentParser(description="Tidal API script")
    parser.add_argument('--country-code', default='US', help='Country code (default: US)')
    parser.add_argument('--get-track', help='Track ID to get details for')
    parser.add_argument('--get-album', help='Album ID to get details for')
    parser.add_argument('--get-track-list', help="Album URL or ID to fetch track list (redirects to --get-album)")
    parser.add_argument('--search-track', help='Track name to search for')
    parser.add_argument('--search-album', help='Album name to search for')
    parser.add_argument('--get-stream', help='Track ID to get stream URL for')
    parser.add_argument('--user-id', help='User ID for stream URL request')
    parser.add_argument('--user-token', help='User access token for stream URL request')
    parser.add_argument('--search-video', help='Video name to search for')
    parser.add_argument('--search-playlist', help='Playlist name to search for')
    parser.add_argument('--search-artist', help='Artist name to search for')

    args = parser.parse_args()

    tidal_api = TidalAPI()

    try:
        if args.get_track:
            track_details = tidal_api.get_track(track_id=args.get_track, country_code=args.country_code)
            print(json.dumps(track_details, indent=4))

        if args.get_album:
            album_details = tidal_api.get_album(album_id=args.get_album, country_code=args.country_code)
            print(json.dumps(album_details, indent=4))

        # Handle --get-track-list as a redirect to --get-album
        if args.get_track_list:
            # Extract the album ID from the input (supports URLs like "album/{albumId}")
            if "album/" in args.get_track_list:
                album_id = args.get_track_list.split("album/")[-1].split("/")[0]
            else:
                album_id = args.get_track_list
            album_details = tidal_api.get_album(album_id=album_id, country_code=args.country_code)
            print(json.dumps(album_details, indent=4))

        if args.search_track:
            search_results = tidal_api.search_tracks(query=args.search_track, country_code=args.country_code)
            print(json.dumps(search_results, indent=4))

        if args.search_album:
            search_results = tidal_api.search_albums(query=args.search_album, country_code=args.country_code)
            print(json.dumps(search_results, indent=4))

        if args.search_video:
            search_results = tidal_api.search_videos(query=args.search_video, country_code=args.country_code)
            print(json.dumps(search_results, indent=4))

        if args.search_playlist:
            search_results = tidal_api.search_playlists(query=args.search_playlist, country_code=args.country_code)
            print(json.dumps(search_results, indent=4))

        if args.search_artist:
            search_results = tidal_api.search_artists(query=args.search_artist, country_code=args.country_code)
            print(json.dumps(search_results, indent=4))

        if args.get_stream:
            stream_url_data = tidal_api.get_stream_url(
                track_id=args.get_stream,
                country_code=args.country_code,
                user_id=args.user_id,
                user_token=args.user_token
            )
            print("Stream URL Data:")
            print(json.dumps(stream_url_data, indent=4))

    except requests.exceptions.RequestException as e:
        print(f"Error: {e}")
        if hasattr(e, 'response') and e.response is not None:
            print(f"Response status code: {e.response.status_code}")
            print(f"Response content: {e.response.content}")

if __name__ == "__main__":
    main()

File: src/funcs/apis/tidalapi.js
const fs   = require('fs');
const path = require('path');

function loadBundledCredentials() {
    try {
        const p = path.join(__dirname, 'apis.json');
        return JSON.parse(fs.readFileSync(p, 'utf8'));
    } catch (_) { return {}; }
}

class TidalAPI {
    constructor(credentials = {}) {
        this.API_URL = "https://openapi.tidal.com/v2";
        this.TOKEN_URL = "https://auth.tidal.com/v1/oauth2/token";
        this.STREAM_API_URL = "https://api.tidal.com/v1";
        this.access_token = null;
        this.token_expiry = null;
        const bundled = loadBundledCredentials();
        const id     = credentials.clientId     || bundled.TIDAL_CLIENT_ID     || '';
        const secret = credentials.clientSecret || bundled.TIDAL_CLIENT_SECRET || '';
        if (!id || !secret) {
            throw new Error("Tidal API credentials not configured. Add them in Settings → API Keys.");
        }
        this.client_id     = id;
        this.client_secret = secret;
    }

    async authenticate() {
        if (this.access_token && this.token_expiry && new Date() < this.token_expiry) {
            return;
        }
        const authHeader = Buffer.from(`${this.client_id}:${this.client_secret}`).toString('base64');
        const headers = {
            "Authorization": `Basic ${authHeader}`,
            "Content-Type": "application/x-www-form-urlencoded"
        };
        const data = "grant_type=client_credentials";
        const response = await fetch(this.TOKEN_URL, {
            method: 'POST',
            headers,
            body: data
        });
        if (!response.ok) {
            throw new Error(`Authentication failed: ${response.statusText}`);
        }
        const tokenData = await response.json();
        this.access_token = tokenData.access_token;
        this.token_expiry = new Date(Date.now() + tokenData.expires_in * 1000);
    }

    async makeRequest(method, url, params = null, baseUrl = null) {
        await this.authenticate();
        const headers = {
            "Authorization": `Bearer ${this.access_token}`,
            "Accept": "application/vnd.tidal.v1+json"
        };
        const fullUrl = `${baseUrl || this.API_URL}/${url}`;
        const query = params ? new URLSearchParams(params).toString() : '';
        const finalUrl = query ? `${fullUrl}?${query}` : fullUrl;
        const response = await fetch(finalUrl, { method, headers });
        if (!response.ok) {
            throw new Error(`Request failed: ${response.statusText}`);
        }
        return await response.json();
    }

    async searchRequest(params) {
        await this.authenticate();

        const { query, type, countryCode, limit } = params;

        const headers = {
            "Authorization": `Bearer ${this.access_token}`,
            "Accept": "application/vnd.api+json"
        };

        const includeTypes = {
            'TRACKS': 'tracks,albums,artists',
            'ALBUMS': 'albums,artists',
            'ARTISTS': 'artists',
            'PLAYLISTS': 'playlists,artists',
            'VIDEOS': 'videos,artists'
        };

        const includeParam = includeTypes[type] || 'tracks';
        const queryParams = new URLSearchParams({
            countryCode,
            include: includeParam
        });

        const finalUrl = `${this.API_URL}/searchResults/${encodeURIComponent(query)}?${queryParams.toString()}`;

        const response = await fetch(finalUrl, { method: 'GET', headers });

        if (!response.ok) {
            const errorText = await response.text();
            throw new Error(`Request failed: ${response.status} ${response.statusText}`);
        }

        const data = await response.json();
        const transformedData = this.transformSearchResponse(data, includeParam);

        return transformedData;
    }

    transformSearchResponse(apiResponse, includeType) {
        const results = apiResponse.included || [];

        const includedMap = {};
        results.forEach(item => {
            if (!includedMap[item.type]) includedMap[item.type] = {};
            includedMap[item.type][item.id] = item;
        });

        const grouped = {
            tracks: [],
            albums: [],
            artists: [],
            playlists: [],
            videos: []
        };

        results.forEach(item => {
            if (item.type === 'tracks') {
                if (item.relationships && item.relationships.albums && item.relationships.albums.data && item.relationships.albums.data.length > 0) {
                    const albumId = item.relationships.albums.data[0].id;
                    const album = includedMap.albums && includedMap.albums[albumId];
                    if (album) {
                        item.album = album.attributes;
                    }
                }
                if (item.relationships && item.relationships.artists && item.relationships.artists.data && item.relationships.artists.data.length > 0) {
                    item.artists = item.relationships.artists.data.map(artistRef => {
                        const artist = includedMap.artists && includedMap.artists[artistRef.id];
                        return artist ? artist.attributes : null;
                    }).filter(Boolean);
                }
                grouped.tracks.push({ resource: item });
            } else if (item.type === 'albums') {
                if (item.relationships && item.relationships.artists && item.relationships.artists.data && item.relationships.artists.data.length > 0) {
                    const artistId = item.relationships.artists.data[0].id;
                    const artist = includedMap.artists && includedMap.artists[artistId];
                    if (artist) {
                        item.artist = artist.attributes;
                    }
                }
                grouped.albums.push({ resource: item });
            } else if (item.type === 'artists') {
                grouped.artists.push({ resource: item });
            } else if (item.type === 'playlists') {
                if (item.relationships && item.relationships.artists && item.relationships.artists.data && item.relationships.artists.data.length > 0) {
                    item.artists = item.relationships.artists.data.map(artistRef => {
                        const artist = includedMap.artists && includedMap.artists[artistRef.id];
                        return artist ? artist.attributes : null;
                    }).filter(Boolean);
                }
                grouped.playlists.push({ resource: item });
            } else if (item.type === 'videos') {
                if (item.relationships && item.relationships.artists && item.relationships.artists.data && item.relationships.artists.data.length > 0) {
                    item.artists = item.relationships.artists.data.map(artistRef => {
                        const artist = includedMap.artists && includedMap.artists[artistRef.id];
                        return artist ? artist.attributes : null;
                    }).filter(Boolean);
                }
                grouped.videos.push({ resource: item });
            }
        });

        return grouped;
    }

    async getStreamUrl(trackId, countryCode, userId = null, userToken = null) {
        const headers = {
            "Authorization": `Bearer ${userToken || this.access_token}`
        };
        const params = { countryCode };
        const url = `${this.STREAM_API_URL}/tracks/${trackId}/streamUrl`;
        const query = new URLSearchParams(params).toString();
        const finalUrl = `${url}?${query}`;
        const response = await fetch(finalUrl, { headers });
        if (!response.ok) {
            throw new Error(`Request failed: ${response.statusText}`);
        }
        return await response.json();
    }

    async getTrack(trackId, countryCode) {
        const params = { countryCode, include: "artists,albums" };
        return await this.makeRequest("GET", `tracks/${trackId}`, params);
    }

    async getAlbum(albumId, countryCode) {
        const albumResponse = await this.makeRequest("GET", `albums/${albumId}`, { countryCode, include: "items" });
        if (albumResponse && !albumResponse.artist) {
            const artistId = albumResponse.artistId;
            if (artistId) {
                const artistResponse = await this.makeRequest("GET", `artists/${artistId}`, { countryCode });
                if (artistResponse) {
                    albumResponse.artistName = artistResponse.name || "Unknown Artist";
                }
            }
        }
        return albumResponse;
    }

    async searchTracks(query, countryCode = 'US', limit = 30) {
        const params = {
            query,
            countryCode,
            limit,
            type: 'TRACKS'
        };
        return await this.searchRequest(params);
    }

    async searchAlbums(query, countryCode = 'US', limit = 30) {
        const params = {
            query,
            countryCode,
            limit,
            type: 'ALBUMS'
        };
        return await this.searchRequest(params);
    }

    async searchArtists(query, countryCode = 'US', limit = 30) {
        const params = {
            query,
            countryCode,
            limit,
            type: 'ARTISTS'
        };
        return await this.searchRequest(params);
    }

    async getArtistAlbums(artistId, countryCode = 'US') {
        const params = { countryCode };
        return await this.makeRequest("GET", `artists/${artistId}/relationships/albums`, params);
    }

    async searchPlaylists(query, countryCode = 'US', limit = 30) {
        const params = {
            query,
            countryCode,
            limit,
            type: 'PLAYLISTS'
        };
        return await this.searchRequest(params);
    }

    async searchVideos(query, countryCode = 'US', limit = 30) {
        const params = {
            query,
            countryCode,
            limit,
            type: 'VIDEOS'
        };
        return await this.searchRequest(params);
    }

    async searchV1(query, type, countryCode, userToken) {
        const typeMap = { track: 'TRACKS', album: 'ALBUMS', artist: 'ARTISTS', playlist: 'PLAYLISTS', video: 'VIDEOS' };
        const tidalType = typeMap[type] || 'TRACKS';
        const url = `https://api.tidal.com/v1/search?query=${encodeURIComponent(query)}&types=${tidalType}&countryCode=${countryCode}&limit=30`;
        const resp = await fetch(url, { headers: { Authorization: `Bearer ${userToken}` } });
        if (!resp.ok) throw new Error(`Tidal v1 search failed: ${resp.status}`);
        const data = await resp.json();
        return {
            tracks:    (data.tracks?.items    || []).map(item => ({ resource: item })),
            albums:    (data.albums?.items    || []).map(item => ({ resource: item })),
            artists:   (data.artists?.items   || []).map(item => ({ resource: item })),
            playlists: (data.playlists?.items || []).map(item => ({ resource: item })),
            videos:    (data.videos?.items    || []).map(item => ({ resource: item })),
        };
    }
}

module.exports = TidalAPI;